File: docs/parser.js
// Stream-parse Apple Health export.xml entirely in the browser.
// Handles .zip (via fflate streaming Unzip) and raw .xml. Memory stays
// bounded throughout: zip bytes are pushed in chunks, fflate inflates
// chunk-by-chunk into a ReadableStream that the line parser consumes
// incrementally. A 1.6 GB inflated XML never sits in memory whole.

const SUPPORTED = [
  { key: "steps",         hk: "HKQuantityTypeIdentifierStepCount",                  label: "步數",            unit: "步",         agg: "sum"  },
  { key: "active_energy", hk: "HKQuantityTypeIdentifierActiveEnergyBurned",         label: "活動消耗",        unit: "kcal",       agg: "sum"  },
  { key: "distance",      hk: "HKQuantityTypeIdentifierDistanceWalkingRunning",     label: "步行距離",        unit: "km",         agg: "sum"  },
  { key: "flights",       hk: "HKQuantityTypeIdentifierFlightsClimbed",             label: "爬樓層",          unit: "層",         agg: "sum"  },
  { key: "hr",            hk: "HKQuantityTypeIdentifierHeartRate",                  label: "平均心率",        unit: "bpm",        agg: "mean" },
  { key: "resting_hr",    hk: "HKQuantityTypeIdentifierRestingHeartRate",           label: "靜息心率",        unit: "bpm",        agg: "mean" },
  { key: "walking_hr",    hk: "HKQuantityTypeIdentifierWalkingHeartRateAverage",    label: "步行心率",        unit: "bpm",        agg: "mean" },
  { key: "hrv",           hk: "HKQuantityTypeIdentifierHeartRateVariabilitySDNN",   label: "HRV",             unit: "ms",         agg: "mean" },
  { key: "spo2",          hk: "HKQuantityTypeIdentifierOxygenSaturation",           label: "血氧 (日均)",     unit: "%",          agg: "mean" },
  { key: "respiratory",   hk: "HKQuantityTypeIdentifierRespiratoryRate",            label: "呼吸頻率",        unit: "次/分",      agg: "mean" },
  { key: "body_temp",     hk: "HKQuantityTypeIdentifierBodyTemperature",            label: "體溫",            unit: "°C",         agg: "mean" },
  { key: "vo2max",        hk: "HKQuantityTypeIdentifierVO2Max",                     label: "VO2 Max",         unit: "ml/kg·min",  agg: "mean" },
  { key: "walking_asymmetry",  hk: "HKQuantityTypeIdentifierWalkingAsymmetryPercentage",     label: "步行不對稱率",  unit: "%",          agg: "mean" },
  { key: "double_support",     hk: "HKQuantityTypeIdentifierWalkingDoubleSupportPercentage", label: "雙腳支撐時間",  unit: "%",          agg: "mean" },
  { key: "daylight",           hk: "HKQuantityTypeIdentifierTimeInDaylight",                 label: "日照時間",      unit: "分鐘",       agg: "sum"  },
  // New in Phase 5 prep — names align with scripts/parse_health.py canonical schema
  { key: "hr_recovery_1min",   hk: "HKQuantityTypeIdentifierHeartRateRecoveryOneMinute",     label: "1 分鐘心率恢復", unit: "bpm",        agg: "mean" },
  { key: "wrist_temp_delta_c", hk: "HKQuantityTypeIdentifierAppleSleepingWristTemperature",  label: "睡眠手腕溫差",  unit: "°C",         agg: "mean" },
  { key: "walking_speed_mps",  hk: "HKQuantityTypeIdentifierWalkingSpeed",                   label: "步行速度",      unit: "m/s",        agg: "mean" },
  { key: "step_length_cm",     hk: "HKQuantityTypeIdentifierWalkingStepLength",              label: "步長",          unit: "cm",         agg: "mean" },
  { key: "walking_steadiness", hk: "HKQuantityTypeIdentifierAppleWalkingSteadiness",         label: "步行穩定度",    unit: "%",          agg: "mean" },
  { key: "six_min_walk_m",     hk: "HKQuantityTypeIdentifierSixMinuteWalkTestDistance",      label: "6 分鐘步行距離", unit: "m",          agg: "mean" },
  { key: "basal_kcal",         hk: "HKQuantityTypeIdentifierBasalEnergyBurned",              label: "基礎代謝",      unit: "kcal",       agg: "sum"  },
  { key: "exercise_minutes",   hk: "HKQuantityTypeIdentifierAppleExerciseTime",              label: "運動時間",      unit: "分鐘",       agg: "sum"  },
  { key: "stand_minutes",      hk: "HKQuantityTypeIdentifierAppleStandTime",                 label: "站立時間",      unit: "分鐘",       agg: "sum"  },
  { key: "headphone_db",       hk: "HKQuantityTypeIdentifierHeadphoneAudioExposure",         label: "耳機音量",      unit: "dB",         agg: "mean" },
  { key: "env_audio_db",       hk: "HKQuantityTypeIdentifierEnvironmentalAudioExposure",     label: "環境音量",      unit: "dB",         agg: "mean" },
  { key: "body_mass_kg",       hk: "HKQuantityTypeIdentifierBodyMass",                       label: "體重",          unit: "kg",         agg: "mean" },
  { key: "bmi",                hk: "HKQuantityTypeIdentifierBodyMassIndex",                  label: "BMI",           unit: "",           agg: "mean" },
];

const SLEEP_HK = "HKCategoryTypeIdentifierSleepAnalysis";
const HK_INDEX = Object.fromEntries(SUPPORTED.map((m) => [m.hk, m]));

export const SUPPORTED_METRICS = SUPPORTED;
export const METRIC_BY_KEY = Object.fromEntries(SUPPORTED.map((m) => [m.key, m]));

// Apple writes dates like "2024-09-12 23:55:21 +0800". JS Date can't parse the
// trailing-offset form reliably, so normalize "YYYY-MM-DD HH:mm:ss +HHMM" -> ISO.
function parseAppleDate(s) {
  if (!s) return null;
  const m = s.match(/^(\d{4})-(\d{2})-(\d{2})[ T](\d{2}):(\d{2}):(\d{2})(?:\s*([+-]\d{2})(\d{2}))?/);
  if (!m) return null;
  const iso = m[7] != null
    ? `${m[1]}-${m[2]}-${m[3]}T${m[4]}:${m[5]}:${m[6]}${m[7]}:${m[8]}`
    : `${m[1]}-${m[2]}-${m[3]}T${m[4]}:${m[5]}:${m[6]}Z`;
  const d = new Date(iso);
  return isNaN(d.getTime()) ? null : d;
}

// Apple decodes XML entities in attribute values; we get them post-decode from
// our regex, but `value` can legitimately contain " in source names. We use a
// non-greedy attribute matcher that handles either "..." or '...'.
const ATTR_RE = /(\w+)\s*=\s*(?:"([^"]*)"|'([^']*)')/g;
function parseAttrs(s) {
  const out = {};
  ATTR_RE.lastIndex = 0;
  let m;
  while ((m = ATTR_RE.exec(s)) !== null) {
    out[m[1]] = m[2] !== undefined ? m[2] : m[3];
  }
  return out;
}

// Match a <Record> opening tag, whether self-closing or wrapping children.
// Apple Health emits self-closing <Record .../> for plain readings, but HRV
// always has <HeartRateVariabilityMetadataList> children, and many SpO2 /
// sleep / heart-rate records carry <MetadataEntry> children — so we accept
// both `/>` and `>`. We only need the attributes from the opening tag; any
// children and the closing </Record> are ignored.
const RECORD_RE = /<Record\b([^>]*?)\/?>/g;

async function* streamLines(stream) {
  // Yields strings broken at newlines; preserves trailing partial buffer.
  const reader = stream.pipeThrough(new TextDecoderStream("utf-8")).getReader();
  let buf = "";
  while (true) {
    const { value, done } = await reader.read();
    if (done) {
      if (buf.length) yield buf;
      return;
    }
    buf += value;
    let nl;
    while ((nl = buf.indexOf("\n")) !== -1) {
      yield buf.slice(0, nl);
      buf = buf.slice(nl + 1);
    }
  }
}

// Apple translates the export filename per device locale.
const KNOWN_EXPORT_NAMES = ["export.xml", "輸出.xml", "导出.xml", "エクスポート.xml"];

// fflate decodes zip entry names as Latin-1 when the spec's UTF-8 flag bit
// (0x800) isn't set, but macOS/iOS-created zips often store UTF-8 bytes in the
// header without setting the flag — the user sees mojibake like
// "apple_health_export/è¼¸å‡º.xml" for what should be "apple_health_export/輸出.xml".
// Recover by re-encoding the string as Latin-1 bytes and decoding them as UTF-8.
function decodeName(name) {
  if (!/[\x80-\xff]/.test(name)) return name;
  try {
    const bytes = new Uint8Array(name.length);
    for (let i = 0; i < name.length; i++) {
      const c = name.charCodeAt(i);
      if (c > 0xff) return name;
      bytes[i] = c;
    }
    return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    return name;
  }
}

function isExportEntry(name) {
  return KNOWN_EXPORT_NAMES.includes(decodeName(name).split("/").pop());
}

// Stream the matching export.xml entry out of a zip without ever holding the
// inflated file in memory. fflate's Unzip is push-based: we feed it zip bytes
// as we read them; for each entry header it sees, onfile fires synchronously.
// We selectively call entry.start() on the matching entry and route its
// inflated chunks into a ReadableStream that the rest of the pipeline reads.
async function getXmlStream(file, onProgress) {
  const head = new Uint8Array(await file.slice(0, 4).arrayBuffer());
  const isZip = head[0] === 0x50 && head[1] === 0x4b;
  if (!isZip) {
    return { stream: file.stream(), totalBytes: file.size };
  }
  if (typeof fflate === "undefined") {
    throw new Error("fflate 未載入，無法解 zip");
  }
  onProgress && onProgress({ phase: "unzip", message: "解壓縮中…" });

  return new Promise((resolveReady, rejectReady) => {
    const seenXmls = [];
    let matched = null;
    let streamController = null;
    let resumePump = null;
    const wakeup = () => { const r = resumePump; resumePump = null; if (r) r(); };

    // 4 MB cap on the inflated queue keeps memory bounded even when fflate
    // outpaces the line parser.
    const xmlStream = new ReadableStream({
      start(c) { streamController = c; },
      pull() { wakeup(); },
    }, new ByteLengthQueuingStrategy({ highWaterMark: 4 * 1024 * 1024 }));

    const unzipper = new fflate.Unzip((entry) => {
      if (entry.name.toLowerCase().endsWith(".xml")) seenXmls.push(decodeName(entry.name));
      if (matched) return;
      if (!isExportEntry(entry.name)) return;
      matched = entry.name;
      entry.ondata = (err, chunk, final) => {
        if (err) { streamController.error(err); return; }
        // chunk may be reused by fflate on the next callback; copy.
        streamController.enqueue(new Uint8Array(chunk));
        if (final) streamController.close();
      };
      entry.start();
      resolveReady({ stream: xmlStream, totalBytes: entry.originalSize || 0 });
    });
    unzipper.register(fflate.UnzipInflate);

    (async () => {
      const reader = file.stream().getReader();
      let bytesRead = 0;
      let lastTick = 0;
      try {
        while (true) {
          const { value, done } = await reader.read();
          if (done) {
            unzipper.push(new Uint8Array(0), true);
            break;
          }
          unzipper.push(value, false);
          bytesRead += value.byteLength;
          // pause until consumer drains below highWaterMark
          while (matched && streamController.desiredSize !== null && streamController.desiredSize <= 0) {
            await new Promise((res) => { resumePump = res; });
          }
          const now = Date.now();
          if (onProgress && !matched && now - lastTick > 100) {
            lastTick = now;
            onProgress({
              phase: "unzip",
              message: "解壓縮中…",
              progress: file.size ? bytesRead / file.size : null,
            });
          }
        }
        if (!matched) {
          rejectReady(new Error(
            `zip 內找不到 export.xml；zip 中的 .xml 檔：${seenXmls.length ? seenXmls.join(", ") : "(無)"}`,
          ));
        }
      } catch (e) {
        if (matched) streamController.error(e);
        else rejectReady(e);
      }
    })();
  });
}

// Date key from a Date in local time (YYYY-MM-DD). Mirrors the helper in
// aggregator.js — used to bucket records into per-day accumulators during
// the streaming-aggregate parse.
function dateKey(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

const PERCENT_KEYS = ["spo2", "walking_asymmetry", "double_support", "walking_steadiness"];

// Streaming-aggregate parser. Instead of accumulating every <Record> into
// arrays and aggregating after parse (which can hit ~100 MB+ for heavy users
// and crash iOS Safari), update a per-day accumulator immediately and throw
// the raw record away. Memory stays in the low MB range regardless of input
// size.
//
// Returns: {
//   dailyAggs:  Map<dateKey, Map<metricKey, { sum, count, min, max, sumSq }>>
//   sleep:      [{ start, end, stage, source }]   // kept full — small set
//   spo2Raw:    [{ start, value }]                // kept full for sleep-min interval join
//   percentMax: { spo2: number, walking_asymmetry: ..., ... }  // for fraction-form detection
//   recordCount: number
// }
export async function parseExport(file, onProgress) {
  const dailyAggs = new Map();
  const sleep = [];
  const spo2Raw = [];
  const percentMax = Object.fromEntries(PERCENT_KEYS.map((k) => [k, -Infinity]));

  function getAcc(dKey, mKey) {
    let dayMap = dailyAggs.get(dKey);
    if (!dayMap) { dayMap = new Map(); dailyAggs.set(dKey, dayMap); }
    let acc = dayMap.get(mKey);
    if (!acc) {
      acc = { sum: 0, count: 0, min: Infinity, max: -Infinity, sumSq: 0 };
      dayMap.set(mKey, acc);
    }
    return acc;
  }

  const { stream, totalBytes } = await getXmlStream(file, onProgress);
  let bytesSeen = 0;
  let recordCount = 0;
  let lastTick = 0;

  for await (const line of streamLines(stream)) {
    bytesSeen += line.length + 1;
    const now = Date.now();
    if (onProgress && now - lastTick > 100) {
      lastTick = now;
      onProgress({
        phase: "parse",
        message: `解析中… ${recordCount.toLocaleString()} 筆紀錄`,
        progress: totalBytes ? bytesSeen / totalBytes : null,
      });
    }
    if (line.indexOf("<Record") === -1) continue;
    RECORD_RE.lastIndex = 0;
    let m;
    while ((m = RECORD_RE.exec(line)) !== null) {
      const attrs = parseAttrs(m[1]);
      const t = attrs.type;
      if (!t) continue;
      const start = parseAppleDate(attrs.startDate);
      const end = parseAppleDate(attrs.endDate);
      if (!start || !end) continue;

      if (t === SLEEP_HK) {
        const stage = (attrs.value || "").replace("HKCategoryValueSleepAnalysis", "");
        sleep.push({ start, end, stage, source: attrs.sourceName || "" });
        recordCount++;
        continue;
      }

      const spec = HK_INDEX[t];
      if (!spec) continue;
      const v = parseFloat(attrs.value);
      if (!Number.isFinite(v)) continue;

      let value = v;
      if (spec.key === "distance") {
        const u = (attrs.unit || "").toLowerCase();
        if (u === "m" || u === "meter" || u === "metre") value = v / 1000;
      }

      // SpO2 raw kept for sleep-window interval join after parse.
      if (spec.key === "spo2") spo2Raw.push({ start, value });

      // Track max for fraction-form percent metrics — used to decide
      // whether to scale ×100 at finalize.
      if (spec.key in percentMax && value > percentMax[spec.key]) {
        percentMax[spec.key] = value;
      }

      // Fold into per-day accumulator and discard the raw record.
      const dKey = dateKey(start);
      const acc = getAcc(dKey, spec.key);
      acc.sum += value;
      acc.count++;
      if (value < acc.min) acc.min = value;
      if (value > acc.max) acc.max = value;
      acc.sumSq += value * value;

      recordCount++;
    }
  }

  // Apply fraction → percent normalization to accumulators in-place.
  // Linear scaling: multiply sum, sumSq scales as ×10000, min/max ×100.
  for (const key of PERCENT_KEYS) {
    const maxV = percentMax[key];
    if (!Number.isFinite(maxV) || maxV > 1.5) continue;
    for (const dayMap of dailyAggs.values()) {
      const acc = dayMap.get(key);
      if (!acc) continue;
      acc.sum *= 100;
      acc.sumSq *= 10000;
      acc.min *= 100;
      acc.max *= 100;
    }
    if (key === "spo2") for (const r of spo2Raw) r.value *= 100;
  }

  sleep.sort((a, b) => a.start - b.start);
  spo2Raw.sort((a, b) => a.start - b.start);

  onProgress && onProgress({
    phase: "done",
    message: `解析完成：${recordCount.toLocaleString()} 筆紀錄`,
    progress: 1,
  });

  return { dailyAggs, sleep, spo2Raw, percentMax, recordCount };
}

File: healthkit/parser.py
"""Stream-parse Apple Health export.xml.

Apple Health exports are often >100MB; we use iterparse and clear elements as we
go so memory stays bounded. Returns one tidy DataFrame per supported metric.
"""
from __future__ import annotations

import io
import zipfile
from dataclasses import dataclass
from pathlib import Path
from typing import IO, Iterable, Mapping
from xml.etree import ElementTree as ET

import pandas as pd
from dateutil import parser as dtparser


@dataclass(frozen=True)
class MetricSpec:
    key: str          # short name used in the dashboard
    hk_type: str      # HealthKit identifier in export.xml
    label_zh: str     # Chinese label for UI
    unit: str         # canonical unit
    aggregation: str  # daily aggregation rule: sum / mean / min / max / sleep


SUPPORTED_TYPES: tuple[MetricSpec, ...] = (
    MetricSpec("steps", "HKQuantityTypeIdentifierStepCount", "步數", "步", "sum"),
    MetricSpec("active_energy", "HKQuantityTypeIdentifierActiveEnergyBurned", "活動消耗", "kcal", "sum"),
    MetricSpec("distance", "HKQuantityTypeIdentifierDistanceWalkingRunning", "步行距離", "km", "sum"),
    MetricSpec("flights", "HKQuantityTypeIdentifierFlightsClimbed", "爬樓層", "層", "sum"),
    MetricSpec("hr", "HKQuantityTypeIdentifierHeartRate", "心率", "bpm", "mean"),
    MetricSpec("resting_hr", "HKQuantityTypeIdentifierRestingHeartRate", "靜息心率", "bpm", "mean"),
    MetricSpec("walking_hr", "HKQuantityTypeIdentifierWalkingHeartRateAverage", "步行心率", "bpm", "mean"),
    MetricSpec("hrv", "HKQuantityTypeIdentifierHeartRateVariabilitySDNN", "心率變異 (HRV)", "ms", "mean"),
    MetricSpec("spo2", "HKQuantityTypeIdentifierOxygenSaturation", "血氧", "%", "mean"),
    MetricSpec("respiratory", "HKQuantityTypeIdentifierRespiratoryRate", "呼吸頻率", "次/分", "mean"),
    MetricSpec("body_temp", "HKQuantityTypeIdentifierBodyTemperature", "體溫", "°C", "mean"),
    MetricSpec("vo2max", "HKQuantityTypeIdentifierVO2Max", "VO2 Max", "ml/kg·min", "mean"),
    MetricSpec("walking_asymmetry", "HKQuantityTypeIdentifierWalkingAsymmetryPercentage", "步行不對稱率", "%", "mean"),
    MetricSpec("double_support", "HKQuantityTypeIdentifierWalkingDoubleSupportPercentage", "雙腳支撐時間", "%", "mean"),
    MetricSpec("daylight", "HKQuantityTypeIdentifierTimeInDaylight", "日照時間", "分鐘", "sum"),
    # New in Phase 5 prep — names align with scripts/parse_health.py canonical schema
    MetricSpec("hr_recovery_1min", "HKQuantityTypeIdentifierHeartRateRecoveryOneMinute", "1 分鐘心率恢復", "bpm", "mean"),
    MetricSpec("wrist_temp_delta_c", "HKQuantityTypeIdentifierAppleSleepingWristTemperature", "睡眠手腕溫差", "°C", "mean"),
    MetricSpec("walking_speed_mps", "HKQuantityTypeIdentifierWalkingSpeed", "步行速度", "m/s", "mean"),
    MetricSpec("step_length_cm", "HKQuantityTypeIdentifierWalkingStepLength", "步長", "cm", "mean"),
    MetricSpec("walking_steadiness", "HKQuantityTypeIdentifierAppleWalkingSteadiness", "步行穩定度", "%", "mean"),
    MetricSpec("six_min_walk_m", "HKQuantityTypeIdentifierSixMinuteWalkTestDistance", "6 分鐘步行距離", "m", "mean"),
    MetricSpec("basal_kcal", "HKQuantityTypeIdentifierBasalEnergyBurned", "基礎代謝", "kcal", "sum"),
    MetricSpec("exercise_minutes", "HKQuantityTypeIdentifierAppleExerciseTime", "運動時間", "分鐘", "sum"),
    MetricSpec("stand_minutes", "HKQuantityTypeIdentifierAppleStandTime", "站立時間", "分鐘", "sum"),
    MetricSpec("headphone_db", "HKQuantityTypeIdentifierHeadphoneAudioExposure", "耳機音量", "dB", "mean"),
    MetricSpec("env_audio_db", "HKQuantityTypeIdentifierEnvironmentalAudioExposure", "環境音量", "dB", "mean"),
    MetricSpec("body_mass_kg", "HKQuantityTypeIdentifierBodyMass", "體重", "kg", "mean"),
    MetricSpec("bmi", "HKQuantityTypeIdentifierBodyMassIndex", "BMI", "", "mean"),
)

SLEEP_TYPE = "HKCategoryTypeIdentifierSleepAnalysis"

_TYPE_INDEX: Mapping[str, MetricSpec] = {m.hk_type: m for m in SUPPORTED_TYPES}


# Apple translates the export filename per device locale, so en uses
# "export.xml" but zh-Hant gives "輸出.xml", zh-Hans "导出.xml", ja "エクスポート.xml".
_KNOWN_EXPORT_NAMES = ("export.xml", "輸出.xml", "导出.xml", "エクスポート.xml")


def _find_export_xml(namelist: list[str]) -> str | None:
    for known in _KNOWN_EXPORT_NAMES:
        for n in namelist:
            if n == known or n.endswith("/" + known):
                return n
    # Fallback: pick a non-CDA .xml sitting in the same directory as export_cda.xml.
    cda = next((n for n in namelist if n.lower().endswith("export_cda.xml")), None)
    prefix = cda.rsplit("/", 1)[0] + "/" if cda and "/" in cda else ""
    for n in namelist:
        ln = n.lower()
        if not ln.endswith(".xml") or ln.endswith("export_cda.xml"):
            continue
        if prefix:
            if n.startswith(prefix) and "/" not in n[len(prefix):]:
                return n
        elif n.count("/") <= 1:
            return n
    return None


def _missing_xml_error(namelist: list[str], where: str) -> ValueError:
    xmls = [n for n in namelist if n.lower().endswith(".xml")] or ["(無 .xml)"]
    return ValueError(f"{where} 內找不到 export.xml；zip 中的 .xml 檔：{xmls}")


def _open_xml(source: str | Path | IO[bytes]) -> IO[bytes]:
    """Accept a path/zip/file-like and return a binary file-like for export.xml."""
    if hasattr(source, "read"):
        head = source.read(4)
        source.seek(0)
        if head[:2] == b"PK":
            zf = zipfile.ZipFile(source)
            name = _find_export_xml(zf.namelist())
            if name is None:
                raise _missing_xml_error(zf.namelist(), "zip")
            return zf.open(name)
        return source  # assume plain XML stream

    path = Path(source)
    if path.suffix.lower() == ".zip":
        zf = zipfile.ZipFile(path)
        name = _find_export_xml(zf.namelist())
        if name is None:
            raise _missing_xml_error(zf.namelist(), str(path))
        return zf.open(name)
    return path.open("rb")


def _to_value(text: str | None) -> float | None:
    if text is None:
        return None
    try:
        return float(text)
    except ValueError:
        return None


def parse_export(source: str | Path | IO[bytes]) -> dict[str, pd.DataFrame]:
    """Parse an Apple Health export and return per-metric tidy DataFrames.

    Result keys: every MetricSpec.key that had data, plus 'sleep' if found.
    Quantity frames have columns: start, end, value, source.
    Sleep frame has columns: start, end, stage, source where stage is one of
    {InBed, Asleep, AsleepCore, AsleepDeep, AsleepREM, Awake}.
    """
    quantity_rows: dict[str, list[tuple]] = {m.key: [] for m in SUPPORTED_TYPES}
    sleep_rows: list[tuple] = []

    stream = _open_xml(source)
    context = ET.iterparse(stream, events=("end",))
    for _, elem in context:
        if elem.tag != "Record":
            continue
        rtype = elem.get("type")
        if rtype == SLEEP_TYPE:
            try:
                start = dtparser.parse(elem.get("startDate"))
                end = dtparser.parse(elem.get("endDate"))
            except (TypeError, ValueError):
                elem.clear()
                continue
            stage = (elem.get("value") or "").replace("HKCategoryValueSleepAnalysis", "")
            sleep_rows.append((start, end, stage, elem.get("sourceName", "")))
        else:
            spec = _TYPE_INDEX.get(rtype)
            if spec is None:
                elem.clear()
                continue
            value = _to_value(elem.get("value"))
            if value is None:
                elem.clear()
                continue
            try:
                start = dtparser.parse(elem.get("startDate"))
                end = dtparser.parse(elem.get("endDate"))
            except (TypeError, ValueError):
                elem.clear()
                continue
            # convert distance metres -> km if Apple emitted m
            unit = elem.get("unit", "")
            if spec.key == "distance" and unit.lower() in {"m", "meter", "metre"}:
                value = value / 1000.0
            quantity_rows[spec.key].append((start, end, value, elem.get("sourceName", "")))
        elem.clear()

    # Apple Health emits these percent metrics as either 0-1 fraction or 0-100
    # percent under unit="%", inconsistently. Normalize to percent: if max
    # observed value is ≤ 1.5, treat as fraction and scale ×100. Without this
    # SpO2 ends up displayed as 0.9 % (would imply death) instead of 90 %.
    percent_keys = {"spo2", "walking_asymmetry", "double_support", "walking_steadiness"}

    out: dict[str, pd.DataFrame] = {}
    for spec in SUPPORTED_TYPES:
        rows = quantity_rows[spec.key]
        if not rows:
            continue
        df = pd.DataFrame(rows, columns=["start", "end", "value", "source"])
        df["start"] = pd.to_datetime(df["start"], utc=True).dt.tz_convert(None)
        df["end"] = pd.to_datetime(df["end"], utc=True).dt.tz_convert(None)
        if spec.key in percent_keys and not df.empty and df["value"].max() <= 1.5:
            df["value"] = df["value"] * 100.0
        out[spec.key] = df.sort_values("start").reset_index(drop=True)

    if sleep_rows:
        sdf = pd.DataFrame(sleep_rows, columns=["start", "end", "stage", "source"])
        sdf["start"] = pd.to_datetime(sdf["start"], utc=True).dt.tz_convert(None)
        sdf["end"] = pd.to_datetime(sdf["end"], utc=True).dt.tz_convert(None)
        out["sleep"] = sdf.sort_values("start").reset_index(drop=True)

    return out


def metric_index() -> dict[str, MetricSpec]:
    return {m.key: m for m in SUPPORTED_TYPES}
